tests/test_main.py

```diff
@@ -30,3 +35,111 @@ def test_set_base64_default_line_size_rejects_negative_values(self):
         with self.assertRaises(ValueError):
             xmlsec.base64_default_line_size(-1)
         self.assertEqual(xmlsec.base64_default_line_size(), size)
+
+
+class TestCallbacks(base.TestMemoryLeaks):
+    def setUp(self):
+        super().setUp()
+        xmlsec.cleanup_callbacks()
+
+    def _sign_doc(self):
+        root = self.load_xml("doc.xml")
+        sign = xmlsec.template.create(
+            root,
+            c14n_method=consts.TransformExclC14N,
+            sign_method=consts.TransformRsaSha1
+        )
+        xmlsec.template.add_reference(
+            sign, consts.TransformSha1, uri="cid:123456")
+
+        ctx = xmlsec.SignatureContext()
+        ctx.key = xmlsec.Key.from_file(
+            self.path("rsakey.pem"), format=consts.KeyDataFormatPem
+        )
+        ctx.sign(sign)
+        return sign
+
+    def _expect_sign_failure(self):
+        exc_info = pytest.raises(xmlsec.Error, self._sign_doc)
+        self.assertEqual(exc_info.value.args, (1, 'failed to sign'))
```

Comment:
Avoid using `pytest.raises` here as it pollutes the memory. Use `unittest` machinery instead, that should fix the issue with stale references: 
```suggestion
        with self.assertRaises(xmlsec.Error) as exc_info:
            self._sign_doc()
        self.assertEqual(exc_info.exception.args, (1, 'failed to sign'))```

Reply:
Or you could even use `with self.assertRaisesRegex(xmlsec.Error, 'failed to sign'): self._sign_doc()`. [Docs](https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertRaisesRegex)

Reply:
Nice, thanks - that was enough to fix the issues locally for me.

Reply:
Looks like it's not - I can also reproduce failing tests on my Fedora 34 using `xmlsec==1.2.29` installed via DNF.

Reply:
Thanks - I had noticed the failures, just haven't had chance to delve into them because of other work commitments. I'll see if I can reproduce them in another environment. Failing that, I'll do some rounds against the CI, so apologies in advance for any noise!